SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipUtils.swift

```diff
@@ -0,0 +1,163 @@
+//===--- OwnershipUtils.swift - Utilities for ownership -------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+// Visited the introducers of a forwarded lifetime.
+//
+// This is used to query lifetime attributes on values, such as "escaping" or "lexical". It must be precise for correctness and is performance critical.
+//
+// TODO: Ideally, this would be defined alongside ForwardingInstruction, but basic data structures and utilities are not available in the SIL module.
+protocol ForwardingUseDefWalker {
+  mutating func introducer(_ value: Value) -> WalkResult
+
+  // Minimally, check a ValueSet. This walker may traverse chains of aggregation and destructuring by default. Implementations may traverse phis.
+  mutating func needWalk(for value: Value) -> Bool
+
+  mutating func walkUp(value: Value) -> WalkResult
+}
+
+extension ForwardingUseDefWalker {
+  mutating func walkUp(value: Value) -> WalkResult {
+    walkUpDefault(value: value)
+  }
+  mutating func walkUpDefault(value: Value) -> WalkResult {
+    if let inst = value.definingInstruction as? ForwardingInstruction
+    {
+      return walkUp(inst: inst)
+    }
+    return introducer(value)
+  }
+  mutating func walkUp(inst: ForwardingInstruction) -> WalkResult {
+    for operand in inst.forwardedOperands {
+      if needWalk(for: operand.value) {
+        if walkUp(value: operand.value) == .abortWalk {
+          return .abortWalk
+        }
+      }
+    }
+    return .continueWalk
+  }
+}
+
+// The underlying value is either an operation that summarizes lifetime attributes, such as a phi or move_value, or it's lifetime has no special attributes and can be freely optimized.
+//
+// TODO: put lifetime queries (escaping, lexical) here.
+struct ForwardingIntroducer {
```

Comment:
Not all values are ForwardingIntroducers. I need a type to enforce that. APIs that directly check lifetime properties without repeating the use-def walk are only correct for the ForwardingIntroducer. 
Removing the type does not remove the abstraction. An abstraction without a type leads to endless confusion and bugs.

Reply:
I see two possibilities to make this more useful (and I guess you also thought of this):
* if the introducer values are just a subset of all `Value` types, we could make a `protocol ForwardingIntroducer : Value`.
* it it's more complicated and additional conditions apply, we could make it an enum like `AccessBase`

Reply:
ForwardingIntroducer is not that useful yet because there's no `isEscaping` or `isLexical` API. I don't need to check it in immediately. It only serves two minor purposes:

1. It prevents anyone from calling `visitForwardedUses` without first asking for the introducer

2. If I have an API that expects to take an introducer as an argument it can use that type. Then it doesn't need to repeat the use-def walk. That's convenient because the walk could give you multiple introducers.